### PR TITLE
enh: support compile time evaluation of implied do loops

### PR DESCRIPTION
**Pull request by Pranavchiku**
Towards #3203

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Here is how LFortran performs against GFortran

```console
% time gfortran e.f90 && time ./a.out
gfortran e.f90  0.67s user 0.05s system 101% cpu 0.712 total
   1.9999999996170159     
./a.out  0.00s user 0.00s system 78% cpu 0.002 total
% time lfortran e.f90 -o e.out && time ./e.out
lfortran e.f90 -o e.out  0.13s user 0.02s system 108% cpu 0.137 total
1.99999999961701591e+00
./e.out  0.00s user 0.00s system 77% cpu 0.003 total
```

**Comment by Pranavchiku** [comment]
MRE for stdlib failure:

```fortran
subroutine test_diag_csp()
complex, parameter :: i_ = cmplx(0,1)
integer :: i
print *, [(i_,i=1,3)]
end subroutine test_diag_csp
```

**Comment by Pranavchiku** [comment]
This works locally for me, although this is not a correct fix, we need to support compile time evaluation for complex types in implied do loop. I think this can be done in subsequent PRs. I'll open new issue and link it.

**Comment by certik** [comment]
Excellent, great job on this!

**Comment by Pranavchiku** [comment]
linked issue: https://github.com/lfortran/lfortran/issues/4022

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -5931,6 +6200,69 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                int itr = 0, curr_nesting_level = 0;
+                // populate compiletime array
+                populate_compiletime_array_for_idl(idl, array, loop_vars, loop_indices, curr_nesting_level, itr);
+
+                data = array;
+            } else if (ASRUtils::is_real(*type)) {
+                int kind = ASRUtils::extract_kind_from_ttype_t(type);
+
+                if (kind == 4) {
+                    float *array = new float[idl_size];
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:6241]: You have to use the allocator here.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -5931,6 +6200,69 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+
+                    int itr = 0, curr_nesting_level = 0;
+                    // populate compiletime array
+                    populate_compiletime_array_for_idl(idl, array, loop_vars, loop_indices, curr_nesting_level, itr);
+
+                    data = array;
+                } else if (kind == 8) {
+                    double *array = new double[idl_size];
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:6249]: same here

**Review by certik — APPROVED** [review]
Later, we can consider moving this to libasr as an "ASR evaluator".

**Comment by certik** [comment]
@Pranavchiku can you post your `--show-asr`? I compiled locally and for me this is giving:
```
                            x:
                                (Variable
                                    2
                                    x
                                    [a
                                    b
                                    i
                                    n]
                                    Local
                                    (ArrayConstructor
                                        [(ImpliedDoLoop
                                            [(IntrinsicElementalFunction
                                                Sin
                                                [(RealBinOp
                                                    (Var 2 a)
                                                    Add
                                                    (RealBinOp
                                                        (RealBinOp
                                                            (RealBinOp
                                                                (Var 2 b)
                                                                Sub
                                                                (Var 2 a)
                                                                (Real 8)
                                                                (RealConstant
                                                                    3.141593
                                                                    (Real 8)
                                                                )
                                                            )
                                                            Mul
                                                            (Cast
                                                                (Var 2 i)
                                                                IntegerToReal
                                                                (Real 8)
                                                                ()
                                                            )
                                                            (Real 8)
                                                            ()
                                                        )
                                                        Div
                                                        (Cast
                                                            (Var 2 n)
                                                            IntegerToReal
                                                            (Real 8)
                                                            (RealConstant
                                                                65535.000000
                                                                (Real 8)
                                                            )
                                                        )
                                                        (Real 8)
                                                        ()
                                                    )
                                                    (Real 8)
                                                    ()
                                                )]
                                                0
                                                (Real 8)
                                                ()
                                            )]
                                            (Var 2 i)
                                            (IntegerConstant 1 (Integer 4))
                                            (Var 2 n)
                                            ()
                                            (Real 8)
                                            ()
                                        )]
                                        (Array
                                            (Real 8)
                                            [((IntegerConstant 1 (Integer 4))
                                            ())]
                                            DescriptorArray
                                        )
                                        ()
                                        ColMajor
                                    )
                                    (ArrayConstructor
                                        [(ImpliedDoLoop
                                            [(IntrinsicElementalFunction
                                                Sin
                                                [(RealBinOp
                                                    (Var 2 a)
                                                    Add
                                                    (RealBinOp
                                                        (RealBinOp
                                                            (RealBinOp
                                                                (Var 2 b)
                                                                Sub
                                                                (Var 2 a)
                                                                (Real 8)
                                                                (RealConstant
                                                                    3.141593
                                                                    (Real 8)
                                                                )
                                                            )
                                                            Mul
                                                            (Cast
                                                                (Var 2 i)
                                                                IntegerToReal
                                                                (Real 8)
                                                                ()
                                                            )
                                                            (Real 8)
                                                            ()
                                                        )
                                                        Div
                                                        (Cast
                                                            (Var 2 n)
                                                            IntegerToReal
                                                            (Real 8)
                                                            (RealConstant
                                                                65535.000000
                                                                (Real 8)
                                                            )
                                                        )
                                                        (Real 8)
                                                        ()
                                                    )
                                                    (Real 8)
                                                    ()
                                                )]
                                                0
                                                (Real 8)
                                                ()
                                            )]
                                            (Var 2 i)
                                            (IntegerConstant 1 (Integer 4))
                                            (Var 2 n)
                                            ()
                                            (Real 8)
                                            ()
                                        )]
                                        (Array
                                            (Real 8)
                                            [((IntegerConstant 1 (Integer 4))
                                            ())]
                                            DescriptorArray
                                        )
                                        ()
                                        ColMajor
                                    )
                                    Parameter
                                    (Array
                                        (Real 8)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 65535 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
``
Which does not seem to have the compile time value set.

**Comment by Pranavchiku** [comment]
Well, I screwed up in last commit when I shifted to `&=`, with #4025, we get:

```clojure
                            x:
                                (Variable
                                    2
                                    x
                                    []
                                    Local
                                    (ArrayConstant
                                        80
                                        [0.3090169943749474, 0.5877852522924731, 0.8090169943749475, ...., 0.5877852522924732, 0.3090169943749475, 0.0000000000000001]
                                        (Array
                                            (Real 8)
                                            [((IntegerConstant 1 (Integer 4))
                                            (IntegerConstant 10 (Integer 4)))]
                                            FixedSizeArray
                                        )
                                        ColMajor
                                    )
                                    (ArrayConstant
                                        80
                                        [0.3090169943749474, 0.5877852522924731, 0.8090169943749475, ...., 0.5877852522924732, 0.3090169943749475, 0.0000000000000001]
                                        (Array
                                            (Real 8)
                                            [((IntegerConstant 1 (Integer 4))
                                            (IntegerConstant 10 (Integer 4)))]
                                            FixedSizeArray
                                        )
                                        ColMajor
                                    )
                                    Parameter
                                    (Array
                                        (Real 8)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 10 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
```

**Comment by certik** [comment]
Yes, that's the result I was expecting!

**Pull request merged**